### PR TITLE
Chapter 11: Account Activation

### PR DESCRIPTION
**Pull request by hazwanizafirah**

### CODE REVIEW

**Comment by hazwanizafirah** [comment]
ready

**Comment by hazwanizafirah** [comment]
Hi. I have a problem, when I try to sign up and click the activation link, it will show that "invalid activation link" and not "account activated". Can I know how to fix this? 
image

File: test/controllers/account_activations_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
```
Comment (anlvq-1058) [inline test/controllers/account_activations_controller_test.rb:2]: You don't need do test please delete this file

**Review by anlvq-1058 — COMMENTED** [review]
cmt

**Comment by hazwanizafirah** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -4,9 +4,15 @@ def new; end
   def create
     @user = User.find_by email: params[:session][:email].downcase
     if @user&.authenticate(params[:session][:password])
+      if @user.activated
       log_in @user
-      params[:session][:remember_me] == "1" ? remember(@user) : forget(@user)
+      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
       redirect_back_or @user
```
Comment (anlvq-1058) [inline app/controllers/sessions_controller.rb:10]: indent, please

File: test/mailers/user_mailer_test.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+require "test_helper"
```
Comment (anlvq-1058) [inline test/mailers/user_mailer_test.rb:1]: delete this file, please

File: test/mailers/previews/user_mailer_preview.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Preview all emails at http://localhost:3000/rails/mailers/user_mailer
+class UserMailerPreview < ActionMailer::Preview
```
Comment (anlvq-1058) [inline test/mailers/previews/user_mailer_preview.rb:2]: delete this file, please

**Review by anlvq-1058 — COMMENTED** [review]
cmt

**Comment by hazwanizafirah** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -4,9 +4,15 @@ def new; end
...
     @user = User.find_by email: params[:session][:email].downcase
     if @user&.authenticate(params[:session][:password])
-      log_in @user
-      params[:session][:remember_me] == "1" ? remember(@user) : forget(@user)
-      redirect_back_or @user
+      if @user.activated
+        log_in @user
+        params[:session][:remember_me] == "1" ? remember(user) : forget(user)
+        redirect_back_or @user
+    else
```
Comment (ngocvt-0484) [inline app/controllers/sessions_controller.rb:11]: wrong indent
it should align with ``if @user.activated``

File: app/models/user.rb [hunk 6]
```diff
@@ -31,11 +33,32 @@ def remember
...
+
+  def activate
+    update_columns activated: true, activated_at: Time.zone.now
+  end
+
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
+  end
+
+private
```
Comment (ngocvt-0484) [inline app/models/user.rb:54]: wrong indent
it should align with `end` above

File: app/views/static_pages/home.html.erb [hunk 7]
```diff
@@ -6,6 +6,7 @@
   <a href="http://www.railstutorial.org/">
   <%= t :p_2 %></a>
   <%= t :p_3 %>
+  <%= link_to t(:sign_up), 'http://127.0.0.1:3000/signup', class: "btn btn-lg btn-primary" %>
```
Comment (ngocvt-0484) [inline app/views/static_pages/home.html.erb:9]:
```suggestion
  <%= link_to t(:sign_up), signup_path, class: "btn btn-lg btn-primary" %>
```

**Comment by hazwanizafirah** [comment]
ready

File: app/controllers/sessions_controller.rb (hunk 2 above)
Comment (anlvq-1058) [reply app/controllers/sessions_controller.rb:10]: Please indent your code like in this picture

**Review by anlvq-1058 — COMMENTED** [review]
cmt

**Comment by hazwanizafirah** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**